include/swift/SIL/SILCloner.h

```diff
@@ -70,48 +102,78 @@ class SILCloner : protected SILInstructionVisitor<ImplClass> {
 
   SILBuilder &getBuilder() { return Builder; }
 
-protected:
-  void beforeVisit(SILInstruction *I) {
-    // Update the set of available opened archetypes with the opened
-    // archetypes used by the current instruction.
-    doPreProcess(I);
+  /// Visit all blocks reachable from the given `StartBB` and all instructions
+  /// in those blocks.
+  ///
+  /// This is used to clone a region within a function and mutates the original
+  /// function. `StartBB` cannot be the function entry block.
+  ///
+  /// The entire CFG is discovered in DFS preorder while cloning non-terminator
+  /// instructions. `visitTerminator` is called in the same order, but only
+  /// after mapping all blocks.
+  void cloneReachableBlocks(SILBasicBlock *startBB,
+                            ArrayRef<SILBasicBlock *> exitBlocks);
+
+  /// Clone all blocks in this function and all instructions in those
+  /// blocks.
+  ///
+  /// This is used to clone an entire function and should not mutate the
+  /// original function.
+  ///
+  /// entryArgs must have a SILValue from the cloned function corresponding to
+  /// each argument in the original function `F`.
+  ///
+  /// Cloned instructions are inserted starting at the end of clonedEntryBB.
+  void cloneFunctionBody(SILFunction *F, SILBasicBlock *clonedEntryBB,
+                         ArrayRef<SILValue> entryArgs);
+
+  /// MARK: Callback utilities used from CRTP extensions during cloning.
+  /// These should only be called from within an instruction cloning visitor.
+
+  /// Visitor callback that registers a cloned instruction. CRTP extensions can
+  /// override the implementation via `postProcess`.
+  void doPostProcess(SILInstruction *Orig, SILInstruction *Cloned) {
+    asImpl().postProcess(Orig, Cloned);
+    assert((Orig->getDebugScope() ? Cloned->getDebugScope() != nullptr : true)
+           && "cloned instruction dropped debug scope");
   }
 
-#define INST(CLASS, PARENT) \
-  void visit##CLASS(CLASS *I);
-#include "swift/SIL/SILNodes.def"
+  /// Visitor callback that maps an original value to an existing value. Called
+  /// whenever the visitor that clones an instruction skips doPostProcess().
+  void foldValue(SILValue origValue, SILValue mappedValue) {
+    ValueMap.insert({origValue, mappedValue});
+  }
```

Comment:
This doesn't seem like the most intuitive name for this method; whenever I see it, I expect it to be a more involved operation, not just updating some internal state.  `setMappedValue`, maybe?

Should this assert that the insertion took effect?  If `origValue` is already mapped, this call is implicitly a no-op, which sounds like a source of bugs.

Reply:
Yes, I added an assert that values aren't remapped multiple times.

I'm not attached to the name, but...
- the SILCloner public API has no notion of a "map" at all.
- during cloning, the visitors have to call one of these two APIs for each original value: `doPostProcess` or `foldValue`. `foldValue` should only be called when there is no new cloned value-producer on which to call `doPostProcess`. This happens whenever the original value-producer is folded during cloning, typically because of type/constant propagation.
- `foldValue` is definitely more meaningful than `doPostProcess` but I'd be happy to rename both APIs if I had a better idea.

I added some comments:
https://github.com/apple/swift/pull/19786/commits/81fa786d6a90c9d1682f319704f3ea5a3b5fef17

Reply:
I see, so it's a more meaningful name when called within the cloner.

I'm not going to insist that you rename this, but maybe `setValueInClone`?

Reply:
The point I'm getting at is that it's important for the cloner implementation *not* to directly map values unless it has folded away an instruction in the cloned code. Otherwise, it must go through the usual `doPostProcess` API, which I'm now calling `recordClonedInstruction`. Hopefully this makes things more clear:

https://github.com/apple/swift/pull/19786/commits/9e440d13a628bd19df23e2646ebc0962b2d1a08c

Reply:
Thanks, that works for me.